lte/gateway/c/oai/tasks/nas/emm/SecurityModeControl.c

```diff
@@ -375,6 +377,46 @@ int emm_proc_security_mode_control(
   OAILOG_FUNC_RETURN(LOG_NAS_EMM, rc);
 }
 
+/****************************************************************************
+ **                                                                        **
+ ** Name:    validate_imei()                                               **
+ **                                                                        **
+ ** Description: Check if the received imei matches with the               **
+ **              blocked imei list                                         **
+ **                                                                        **
+ ** Inputs:  imei/imeisv string : imei received in security mode           **
+ **                               complete/attach req                      **
+ ** Outputs:                                                               **
+ **      Return:    EMM cause                                              **
+ **      Others:    None                                                   **
+ **                                                                        **
+ ***************************************************************************/
+int validate_imei(imeisv_t* imeisv) {
+  OAILOG_FUNC_IN(LOG_NAS_EMM);
+  /* First convert only TAC to uint64_t. If TAC is not found in the hashlist,
+   * convert IMEI(TAC + SNR) into uint64_t and check if the key is found
+   * the hashlist
+   */
+  imei64_t tac64 = 0;
+  IMEI_MOBID_TO_IMEI_TAC64(imeisv, &tac64);
+  hashtable_rc_t h_rc = hashtable_uint64_ts_is_key_exists(
+      mme_config.blocked_imei.imei_htbl, (const hash_key_t) tac64);
```

Comment:
if there are no IMEIs in the blocked list, we are making this expensive computation. first check if hash table has any entries, if none return success otherwise proceed with the checks.